사용자 알림 목록 조회

Kal-MH opened the pull request:
# 이슈 번호가 있다면 적어주세요
[#106]

## 작업 목록

- [x]  알림 사용자 조회
- [x]  컴포넌트 스타일링
- [x]  알림 아이콘 뱃지

### 참고 사진이 있다면 첨부
![화면 캡처 2023-01-16 154127](https://user-images.githubusercontent.com/59648372/212615829-5692cb4b-34c0-4125-ac42-8dd8abe8ce9a.png)

## 예정

1. 알림 확인
2. 알림 삭제
3. 알림 무한 스크롤

## 궁금한 점

Kal-MH committed:
feat: get alarm list

Kal-MH committed:
feat: 불러온 알림 목록 Notification List 컴포넌트에 주입

Kal-MH committed:
feat: notification button badge 추가

Kal-MH committed:
feat: alarm 시간 보여주기

Kal-MH committed:
design: like, comment, follow, message에 해당하는 아이콘 추가

Kal-MH committed:
refactor: calcElapsedTime hook으로 분리

Kal-MH committed:
Merge branch 'develop' into feature/alarm-list

Kal-MH committed:
fix & refactor: like, comment author id, useEffect dependency (tabValue)

Kal-MH committed:
Merge branch 'develop' into feature/alarm-list

glassk commented on src/hooks/useComment.ts:
```diff
@@ -47,7 +51,18 @@ export const useCommentForm = () => {
       }
-      const { _id, author, post } = await postStoryComment(comment, storyId);
+
+      /*
+       * postStoryComment의 author -> comment를 작성한 사람(나)
+       * getStoriesDetail -> 게시글을 통해서 게시글 작성자 아이디 얻기
+       * postNotification("COMMENT", _id, 게시글 작성자 아이디, 게시글 아이디);
+       */
```
확인했습니당! ✅

glassk commented on src/components/Story/StoryInfo.tsx:
```diff
@@ -81,7 +81,7 @@ const StoryInfo = ({ story }: Props) => {
         {!isLoading && user && (
           <LikeButton
-            userId={user._id}
+            userId={story.author._id}
```
✅

glassk commented on src/apis/notification.ts:
```diff
@@ -1,6 +1,22 @@
+
+    return notificationList;
+  } catch (error) {
+    //Thinking: console.error 보여줘야 하는가?
+    // console.error(error);
```
일단은 다른 곳처럼 보여주면 좋을 것 같아요!

metacode22 replied:
console.error 일단 찍어주고 나중에 지워주는 건 어떨까요?

glassk commented on src/hooks/useCalcElaspedTime.ts:
```diff
@@ -0,0 +1,24 @@
+
+    const elapsedTime = curTime.getTime() - createdTime.getTime();
+
+    const eDay = Math.floor(elapsedTime / (1000 * 60 * 60 * 24));
+    const eHour = Math.floor(elapsedTime / (1000 * 60 * 60));
+    const eMinutes = Math.floor(elapsedTime / (1000 * 60));
+
+    if (eDay === 0) {
+      if (eHour === 0) return `${eMinutes}분 전`;
+      return `${eHour}시간 전`;
+    }
+    return `${eDay}일 전`;
```
세심하게 처리해주셨네요 👍

glassk commented on src/components/Alarm/NotificationMsg.tsx:
```diff
@@ -1,27 +1,106 @@
+  const navigate = useNavigate();
+  const { calcCurrentToCreatedDate } = useCalcElaspedTime();
+
+  const generateMsg = () => {
+    if (like) return `${fullName}${LIKE}`;
+    if (comment) return `${fullName}${COMMENT}`;
+    if (follow) return `${fullName}${FOLLOW}`;
+    if (message) return `${fullName}${MESSAGE}`;
+  };
+
+  const generateAvatar = () => {
+    if (like) return <ThumbUp />;
+    if (comment) return <CommentIcon />;
+    if (follow) return <Avatar alt={fullName} src={image ? image : ''} />;
+    if (message) return <SendIcon />;
+  };
```
넘 좋습니다

glassk reviewed (approved):
작업량이 꽤 많은데 고생 많으셨어요! 제가 놓친 부분까지 캐치해주셔서 감사해요😃

metacode22 commented on src/components/Alarm/NotificationButton.tsx:
```diff
@@ -18,10 +25,13 @@ const AlarmButton = () => {
-    <Box>
-      <Notifications onClick={handleClick} />
+    <Box sx={{ cursor: 'pointer' }}>
+      {/* thinking: badgeContent={4}를 통해서 안에 숫자를 넣어주는 방향도 있다. */}
```
좋아유! 알림 몇 개가 왔는지 확인할 수 있으면 좋을 것 같습니다!

Kal-MH replied:
일단 넣어봤는데 뱃지 크기가 아이콘에 비해서 좀 크더라구요 다음 PR때 사진을 한 번 올려보겠습니다:)

metacode22 commented on src/components/Alarm/NotificationList.tsx:
```diff
@@ -1,44 +1,49 @@
 import { List } from '@mui/material';
+import { useEffect, useState } from 'react';
 
+import { Notification } from '../../interfaces/noti';
```
파일이름 notification으로 풀어쓰는 거에 대해서는 어떻게 생각하시나요?

Kal-MH replied:
이게, 파일이름이 겹치는 거 같아서 이렇게 했는데, 생각해보니 상관 없겠네요 흐

metacode22 commented on src/components/Alarm/NotificationMsg.tsx:
```diff
@@ -1,27 +1,106 @@
+  FOLLOW: `님이 회원님을 팔로우합니다. `,
+  MESSAGE: `님이 회원님께 메세지를 보냈습니다.`,
+  COMMENT: `님이 회원님 게시글에 댓글을 남겼습니다.`,
+};
```
👍!!!

metacode22 commented on src/hooks/useCalcElaspedTime.ts:
```diff
@@ -0,0 +1,24 @@
+const useCalcElaspedTime = () => {
```
react hook이 사용되지 않는 훅이라서 따로 use로 안빼도 되지 않을까 싶어요🥲 밑의 사진을 보면 use로 쓰면 해당 커스텀 훅이 hook rules에 위배되는지 검사하게 된다고 나오네요. use로 쓰면 커스텀 훅으로 인식한다는 건데, react hook이 들어가지 않는 로직을 커스텀 훅으로 작성하면 혼동이 생길 수 있지 않을까 염려가 돼서요! 저희끼리도 이건 커스텀 훅이구나 인식하고 나중에 렌더링을 신경쓰거나 할 때 조금은 오해의 여지가 생길 수 있을 것 같습니다.

![image](https://user-images.githubusercontent.com/93233930/212619966-ed4ae193-bd69-4a82-8311-235f325ea268.png)

이건 제 추측입니다!
이전에 reference 두 번째 링크의 영상을 본 적이 있는데, react hook이 특정 공간을 차지하는 것 같아요. 근데 거기에 (물론 hook rules에 위배되면 들어가진 않을 것 같습니다) 필요없는 것이 들어가면 낭비가 되거나 나중에 hook이 자신의 값이 저장된 위치를 찾아야 하는데(이건 글로 설명하기 힘드네요...) 못 찾게 되어 오류가 생길 수 있지 않을까...라는 생각이 듭니다.

TMI 죄송해요 ㅋㅋㅋ...

**reference**
https://reactjs.org/docs/hooks-custom.html#using-a-custom-hook
[Can Swyx recreate React Hooks and useState in under 30 min? - JSConf.Asia](https://www.youtube.com/watch?v=KJP1E-Y-xyo)

Kal-MH replied:
아니요, 이런 피드백 너무 귀하죠 ㅎㅎ

처음에 useNotification으로 notification 상태를 관리하는 훅으로 분리했다가 그게 잘 안돼서 calcCurrentToCreatedTime 함수만 남았습니다..

이런 부분이 있었네요:)! 공유해주셔서 감사합니다 
urils로 따로 분리해보도록 할께요:)

glassk replied:
넘 좋은 내용이네요 한번 더 짚고 갑니다!👍 공유 감사합니다~

metacode22 reviewed (commented):
useCalcElaspedTime을 hook으로 뺀 이유가 따로 있을까유? react hook을 사용하지 않는다면 utils로 빼면 될 것 같아서요! 고생하셨습니다😀 구웃~~ LGTM!!!

Kal-MH committed:
refactor:
1. noti -> notification 파일 이름 변경
2. useCalcElapsedTime -> calcCreatedToCurrenttime 변경 및 utils로 이동

tooooo1 commented on src/components/Alarm/NotificationButton.tsx:
```diff
@@ -8,7 +8,14 @@ import { getLocalStorage } from '../../utils/storage';
 const { NOTIFICATION, SIGNIN } = ROUTES;
 
-const AlarmButton = () => {
+/*
+ * TODO: 알림 확인하는 방법
+ *  1. 주기적으로 알림 체크하기?
+ *  2. 확인 안한 알림 있으면 뱃지로 표시하기
+ *  3. 알림이 오는 순간, 알림페이지에 있다면 -> 확인, 미확인을 나누긴 해야할 듯
+ */
```
api 확인해보니 알림 읽음 api도 있네요

2번이랑 3번 둘 다 적용할 수 있을 것 같은데 하면 좋을 거 같아요!

tooooo1 commented on src/components/Alarm/NotificationList.tsx:
```diff
@@ -1,44 +1,49 @@
-        : dummyPostArr.map((m, i) => <NotificationMsg key={i} message={m} />)}
+      {type === MESSAGE
+        ? msgList?.map((n, i) => <NotificationMsg key={i} noti={n} />)
+        : postList?.map((n, i) => <NotificationMsg key={i} noti={n} />)}
```
혹시 noti는 머예용?

tooooo1 commented on src/components/Alarm/NotificationMsg.tsx:
```diff
@@ -1,27 +1,105 @@
 
 interface Props {
-  message: string;
+  noti: Notification;
```
아하 여기있네요

tooooo1 commented on src/components/Alarm/NotificationMsg.tsx:
```diff
@@ -1,27 +1,105 @@
+
+const { STORY_BOOK_BY_USER_ID, STORY_BY_STORY_ID } = ROUTES;
+
+//TODO
+//알림 확인할 때, 색상 변화값 주기 or 뱃지 사라지게 하기
```
뱃지는 mui에서 숫자 체킹도 해줄 거라서 별다른 비용은 안 들 것 같아요. 색상 변화값 좋은거같아융!

tooooo1 commented on src/pages/Notification.tsx:
```diff
@@ -67,7 +78,7 @@ const Notification = () => {
         </Box>
         <Box component='section'>
-          <NotificationList type={tabValue} />
+          <NotificationList type={tabValue} notifications={notifications} />
```
Msg랑 통일하면 좋을 것 같아요!

tooooo1 reviewed (approved):
작업량 무엇..! 대박대박

Kal-MH committed:
refactor: noti - notification 변수명 변경
